File: firecrawl-mcp-server/filter_strains.py
import pandas as pd
import numpy as np
import os

def filter_incomplete_strains(input_file, output_file, min_fields=3):
    """
    Filter out incomplete strain entries from a CSV file.
    
    Parameters:
    - input_file: Path to the input CSV file
    - output_file: Path to save the filtered CSV 
    - min_fields: Minimum number of non-empty fields required in critical columns
    """
    print(f"Reading data from: {input_file}")
    
    # Read the CSV file
    df = pd.read_csv(input_file)
    
    # Count total strains before filtering
    total_strains = len(df)
    print(f"Total strains before filtering: {total_strains}")
    
    # Critical columns that should have values
    critical_columns = [
        'thc_percent', 'average_rating', 'rating_count', 
        'effects', 'medical', 'flavors', 'description'
    ]
    
    # Handle NaN values in string columns
    string_columns = ['effects', 'medical', 'flavors', 'description']
    for col in string_columns:
        df[col] = df[col].fillna('')
    
    # Count non-empty values in critical columns
    df['critical_field_count'] = 0
    for col in critical_columns:
        if col in string_columns:
            df['critical_field_count'] += (df[col] != '').astype(int)
        else:
            df['critical_field_count'] += df[col].notna().astype(int)
    
    # Create a boolean mask for filtering
    mask = (
        # Must have THC value OR CBD value
        (df['thc_percent'].notna() | df['cbd_percent'].notna()) &
        # Must have at least some effects, medical uses, or flavors
        ((df['effects'] != '') | (df['medical'] != '') | (df['flavors'] != '')) &
        # Must have a description
        (df['description'] != '') &
        # Must have at least min_fields critical fields with values
        (df['critical_field_count'] >= min_fields)
    )
    
    # Apply the filter
    filtered_df = df[mask].drop(columns=['critical_field_count'])
    
    # Count total strains after filtering
    filtered_count = len(filtered_df)
    removed_count = total_strains - filtered_count
    
    print(f"Strains kept: {filtered_count} ({filtered_count/total_strains:.1%})")
    print(f"Strains removed: {removed_count} ({removed_count/total_strains:.1%})")
    
    # Save the filtered data
    filtered_df.to_csv(output_file, index=False)
    print(f"Filtered data saved to {output_file}")
    
    # Generate detailed statistics about remaining data
    stats = {
        'total_strains': total_strains,
        'kept_strains': filtered_count,
        'removed_strains': removed_count,
        'kept_percentage': filtered_count/total_strains,
        'examples_removed': df[~mask]['name'].head(10).tolist(),
        'field_completeness': {}
    }
    
    # Calculate completeness of each field in the filtered dataset
    for column in df.columns:
        if column == 'critical_field_count':
            continue
            
        if column in string_columns:
            non_empty = (filtered_df[column] != '').sum()
        else:
            non_empty = filtered_df[column].notna().sum()
            
        stats['field_completeness'][column] = {
            'count': non_empty,
            'percentage': non_empty / filtered_count
        }
    
    return stats

# Path configuration
input_file = 'processed-data/standardized-strains.csv'
output_file = 'processed-data/filtered-strains.csv'
stats_file = 'processed-data/filtering-stats.txt'

# Example usage
if __name__ == "__main__":
    # Make sure output directory exists
    os.makedirs(os.path.dirname(output_file), exist_ok=True)
    
    # Run filtering
    print("\n🔍 FILTERING INCOMPLETE STRAIN DATA")
    print("===================================")
    stats = filter_incomplete_strains(input_file, output_file, min_fields=3)
    
    # Output more detailed statistics
    print("\n📊 FIELD COMPLETENESS IN FILTERED DATASET")
    print("=======================================")
    for field, values in stats['field_completeness'].items():
        print(f"{field}: {values['count']} records ({values['percentage']:.1%})")
    
    print("\n❌ EXAMPLES OF REMOVED STRAINS")
    print("============================")
    for strain in stats['examples_removed']:
        print(f"- {strain}")
    
    # Save statistics to file
    with open(stats_file, 'w') as f:
        f.write(f"STRAIN DATA FILTERING STATISTICS\n")
        f.write(f"==============================\n\n")
        f.write(f"Total strains before filtering: {stats['total_strains']}\n")
        f.write(f"Strains kept: {stats['kept_strains']} ({stats['kept_percentage']:.1%})\n")
        f.write(f"Strains removed: {stats['removed_strains']} ({1-stats['kept_percentage']:.1%})\n\n")
        
        f.write(f"FIELD COMPLETENESS IN FILTERED DATASET\n")
        f.write(f"-------------------------------------\n")
        for field, values in stats['field_completeness'].items():
            f.write(f"{field}: {values['count']} records ({values['percentage']:.1%})\n")
        
        f.write(f"\nEXAMPLES OF REMOVED STRAINS\n")
        f.write(f"---------------------------\n")
        for strain in stats['examples_removed']:
            f.write(f"- {strain}\n")
    
    print(f"\n✅ Complete filtering statistics saved to {stats_file}")
    print(f"✅ Filtered data saved to {output_file}") 

File: firecrawl-mcp-server/filter-strains.js
import fs from 'fs/promises';
import { createReadStream } from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { createInterface } from 'readline';

// Get directory name equivalent in ESM
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Filter incomplete strain data entries based on specified criteria
 */
async function filterIncompleteStrains() {
  console.log('🧹 Starting strain data filtering...');
  
  // Configuration
  const INPUT_FILE = path.join(__dirname, 'processed-data/standardized-strains.json');
  const OUTPUT_FILE = path.join(__dirname, 'processed-data/filtered-strains.json');
  const OUTPUT_CSV = path.join(__dirname, 'processed-data/filtered-strains.csv');
  const STATS_FILE = path.join(__dirname, 'processed-data/filtering-stats.json');
  
  // Minimum number of critical fields required
  const MIN_CRITICAL_FIELDS = 3;
  
  // Critical columns that should have values
  const CRITICAL_COLUMNS = [
    'thc_percent', 'average_rating', 'rating_count', 
    'effects', 'medical', 'flavors', 'description'
  ];
  
  try {
    // Read the JSON file
    console.log(`Reading data from: ${INPUT_FILE}`);
    const data = JSON.parse(await fs.readFile(INPUT_FILE, 'utf8'));
    
    // Count total strains before filtering
    const totalStrains = data.length;
    console.log(`Total strains before filtering: ${totalStrains}`);
    
    // Apply filtering
    const filteredStrains = data.filter(strain => {
      // Count non-empty critical fields
      let criticalFieldCount = 0;
      
      for (const field of CRITICAL_COLUMNS) {
        if (field === 'effects' || field === 'medical' || field === 'flavors' || field === 'description') {
          // Array fields should have length > 0 or string should not be empty
          if (Array.isArray(strain[field]) && strain[field].length > 0) {
            criticalFieldCount++;
          } else if (typeof strain[field] === 'string' && strain[field].trim() !== '') {
            criticalFieldCount++;
          }
        } else {
          // Numeric fields should not be null
          if (strain[field] !== null && strain[field] !== undefined) {
            criticalFieldCount++;
          }
        }
      }
      
      // Apply filtering criteria
      return (
        // Must have THC value OR CBD value
        ((strain.thc_percent !== null && strain.thc_percent !== undefined) || 
         (strain.cbd_percent !== null && strain.cbd_percent !== undefined)) &&
        // Must have at least some effects, medical uses, or flavors
        ((Array.isArray(strain.effects) && strain.effects.length > 0) || 
         (Array.isArray(strain.medical) && strain.medical.length > 0) || 
         (Array.isArray(strain.flavors) && strain.flavors.length > 0)) &&
        // Must have a description
        (typeof strain.description === 'string' && strain.description.trim() !== '') &&
        // Must have at least MIN_CRITICAL_FIELDS critical fields with values
        (criticalFieldCount >= MIN_CRITICAL_FIELDS)
      );
    });
    
    // Count total strains after filtering
    const filteredCount = filteredStrains.length;
    const removedCount = totalStrains - filteredCount;
    
    console.log(`Strains kept: ${filteredCount} (${(filteredCount/totalStrains*100).toFixed(1)}%)`);
    console.log(`Strains removed: ${removedCount} (${(removedCount/totalStrains*100).toFixed(1)}%)`);
    
    // Calculate field completeness for the filtered dataset
    const fieldCompleteness = {};
    
    for (const field of ['name', 'classification', 'thc_percent', 'cbd_percent', 'cbg_percent', 
                         'average_rating', 'rating_count', 'effects', 'medical', 'flavors', 
                         'terpenes', 'lineage', 'awards', 'description']) {
      
      let count = 0;
      
      for (const strain of filteredStrains) {
        if (Array.isArray(strain[field])) {
          if (strain[field].length > 0) count++;
        } else if (typeof strain[field] === 'string') {
          if (strain[field].trim() !== '') count++;
        } else {
          if (strain[field] !== null && strain[field] !== undefined) count++;
        }
      }
      
      fieldCompleteness[field] = {
        count,
        percentage: count / filteredCount
      };
    }
    
    // Examples of removed strains
    const removedStrains = data
      .filter(strain => !filteredStrains.includes(strain))
      .slice(0, 10)
      .map(strain => strain.name);
    
    // Generate statistics object
    const stats = {
      timestamp: new Date().toISOString(),
      total_strains: totalStrains,
      kept_strains: filteredCount,
      removed_strains: removedCount,
      kept_percentage: filteredCount / totalStrains,
      examples_removed: removedStrains,
      field_completeness: fieldCompleteness
    };
    
    // Save filtered data
    console.log(`Saving filtered data to ${OUTPUT_FILE}...`);
    await fs.writeFile(OUTPUT_FILE, JSON.stringify(filteredStrains, null, 2));
    
    // Generate CSV
    console.log(`Converting to CSV format...`);
    const csv = generateCSV(filteredStrains);
    await fs.writeFile(OUTPUT_CSV, csv);
    
    // Save statistics
    console.log(`Saving filtering statistics to ${STATS_FILE}...`);
    await fs.writeFile(STATS_FILE, JSON.stringify(stats, null, 2));
    
    // Output filtering summary
    console.log('\n📊 FIELD COMPLETENESS IN FILTERED DATASET');
    console.log('=======================================');
    for (const [field, values] of Object.entries(fieldCompleteness)) {
      console.log(`${field}: ${values.count} records (${(values.percentage*100).toFixed(1)}%)`);
    }
    
    console.log('\n❌ EXAMPLES OF REMOVED STRAINS');
    console.log('============================');
    for (const strain of removedStrains) {
      console.log(`- ${strain}`);
    }
    
    console.log(`\n✅ Filtering complete! Kept ${filteredCount}/${totalStrains} strains (${(filteredCount/totalStrains*100).toFixed(1)}%)`);
    
    return stats;
    
  } catch (error) {
    console.error(`Error during strain filtering: ${error.message}`);
    throw error;
  }
}

/**
 * Generates CSV content from strain data
 */
function generateCSV(strains) {
  // Define headers (flat structure)
  const headers = [
    'name', 'classification', 'thc_percent', 'cbd_percent', 'cbg_percent',
    'average_rating', 'rating_count', 'effects', 'medical', 'flavors', 'terpenes',
    'lineage', 'awards', 'grow_info_yield', 'grow_info_height', 'grow_info_difficulty',
    'grow_info_flowering_time_weeks', 'description', 'url'
  ];
  
  // Create CSV header row
  let csv = headers.join(',') + '\n';
  
  // Add each strain as a row
  for (const strain of strains) {
    const row = headers.map(header => {
      if (header.includes('grow_info_')) {
        // Handle nested grow_info
        const subField = header.replace('grow_info_', '');
        const value = strain.grow_info ? strain.grow_info[subField] : '';
        return formatCSVField(value);
      } else if (['effects', 'medical', 'flavors', 'terpenes', 'lineage', 'awards'].includes(header)) {
        // Join arrays with semicolons
        const value = Array.isArray(strain[header]) ? strain[header].join(';') : '';
        return formatCSVField(value);
      } else {
        // Regular fields
        return formatCSVField(strain[header]);
      }
    });
    
    csv += row.join(',') + '\n';
  }
  
  return csv;
}

/**
 * Formats a field for CSV output
 */
function formatCSVField(value) {
  if (value === null || value === undefined) return '';
  
  // Convert to string and escape quotes
  const str = String(value).replace(/"/g, '""');
  
  // Wrap in quotes if contains comma, newline or quote
  if (str.includes(',') || str.includes('\n') || str.includes('"') || str.includes(';')) {
    return `"${str}"`;
  }
  
  return str;
}

// Run the filtering
filterIncompleteStrains().catch(error => {
  console.error('Error during strain filtering:', error);
  process.exit(1);
}); 